src/class_system.js

```diff
@@ -0,0 +1,273 @@
+(function(o) {
+    var info = {};
+    var protocols = {};
+
+    var classTip = ' Use: class MyClass extends BaseClass {...}';
+
+    function cloneProperties(source, target) {
+        Object.keys(source).forEach(function(propName){
+            target[propName] = source[propName];
+        });
+        return target;
+    }
+
+    function checkClassName(className) {
+        if (typeof className !== 'string' || !className.length) {
+            throw new Error('className name must be string and of length > 0.');
+        }
+    }
+
+    function defineCheck(className, classConstructor) {
+        checkClassName(className);
+        if (typeof classConstructor !== 'function') {
+            throw new Error('Class constructor must be function.');
+        }
+        var classInfo = ensureClassInfo(className);
+        if (classInfo.classConstructor) {
+            throw new Error('Class "' + className + '" already defined. Use Oskari.clazz.category(...) to add methods to existing class.');
+        }
+        return classInfo;
+    }
+
+    function checkFinalized(classInfo) {
+        if (!classInfo.finalized) {
+            throw new Error('Definitions missing for "' + className + '"! Make sure class and its super classes are loaded.');
+        }
+    }
+
+    function subClassingCheck(classInfo) {
+        if (classInfo.superClass && !classInfo.superClass.allowSubClass) {
+            throw new Error('Not possible to extend ES class via metadata. Check class "' + classInfo.className + '".' + classTip);
+        }
+    }
+
+    function getClassInfo(className) {
+        checkClassName(className)
+        var classInfo = info[className];
+        if (!classInfo) {
+            throw new Error('No class registered with name: ' + className);
+        }
+        return classInfo;
+    }
+
+    function ensureClassInfo(className) {
+        var classInfo = info[className];
+        if (!classInfo) {
+            classInfo = {
+                className: className,
+                classConstructor: null,
+                classPrototype: {},
+                metadata: null,
+                superClass: null,
+                subClasses: [],
+                allowSubClass: true, // ES6 classes cannot be subclassed with ES5 constructors
+                finalized: false // is ready to be instantiated? true, when all classes in super class chain have been registered
+            }
+            info[className] = classInfo;
+        }
+        return classInfo;
+    }
+
+    function processMetadata(classInfo, metadata) {
+        if (!metadata) {
+            return true;
+        }
+        classInfo.metadata = metadata;
+        processProtocols(classInfo, metadata.protocol);
+        if (!metadata.extend) {
+            return true;
+        }
+        if (!Array.isArray(metadata.extend) || metadata.extend.length > 1) {
```

Comment:
Doesn't really make sense to require an array with only one value. Looking at any existing code will make you think that adding classes to the array will work. But as they are currently arrays in the existing code it's a minor grievance since we are moving towards the ES6 class syntax anyway. Nice that it throws an error with a clear message.

Reply:
Yes, now that only single inheritance is supported, the array is no longer needed other than for backward compatibility with existing code.